charm/kafka/reactive/tls.py

```diff
@@ -163,3 +163,63 @@ def import_ca_crt_to_keystore():
 
             remove_state('tls_client.ca_installed')
             set_state('kafka.ca.keystore.saved')
+
+
+@when('endpoint.certificates.departed')
+def clear_certificates():
+    ca_path = '/usr/local/share/ca-certificates/{}.crt'.format(
+        hookenv.service_name()
+    )
+    ca_keystore = os.path.join(
+        KAFKA_SNAP_DATA,
+        "kafka.server.truststore.jks"
+    )
+    if os.path.exists(ca_path):
+        os.remove(ca_path)
+    if os.path.exists(ca_keystore):
+        os.remove(ca_keystore)
+
+    certs_paths = unitdata.kv().get('layer.tls-client.cert-paths', {})
+    for cert_type in ('server', 'client'):
+        crt_path = os.path.join(
+            KAFKA_SNAP_DATA,
+            "{}.crt".format(cert_type)
+        )
+        key_path = os.path.join(
+            KAFKA_SNAP_DATA,
+            "{}.key".format(cert_type)
+        )
+        keystore_path = os.path.join(
+            KAFKA_SNAP_DATA,
+            "kafka.{}.jks".format(cert_type)
+        )
+        if os.path.exists(crt_path):
+            os.remove(crt_path)
+        if os.path.exists(key_path):
+            os.remove(key_path)
+        if os.path.exists(keystore_path):
+            os.remove(keystore_path)
+        remove_state('kafka.{}.keystore.saved'.format(cert_type))
+        data_changed('kafka_{}_certificate'.format(cert_type), {})
+
+        for common_name, paths in certs_paths.get(cert_type, {}).items():
+            data_changed('layer.tls-client.'
+                         '{}.{}'.format(cert_type, common_name), {})
+
+    unitdata.kv().unsetrange(None, 'reactive.tls_client')
```

Comment:
Let's add a comment indicating why this is necessary -- it's to compensate for the underlying layer failing to clear flags and data_changed state properly.

We might want to open a bug on that layer as well and reference this workaround since CDK also uses it.

Reply:
added a comment